geom/twkb_parser.go

```diff
@@ -0,0 +1,562 @@
+	return TWKBParser{
+		twkb:       twkb,
+		opts:       opts,
+		ctype:      DimXY,
+		dimensions: 2,
+	}
+}
+
+// Parse a geometry and return it, the number of bytes consumed, and any error.
+func (p *TWKBParser) parseGeometry() (Geometry, int, error) {
+	geom, err := p.nextGeometry()
+	return geom, p.pos, err
+}
+
+// Parse a geometry and return it and any error.
+func (p *TWKBParser) nextGeometry() (Geometry, error) {
+	if err := p.parseTypeAndPrecision(); err != nil {
+		return Geometry{}, err
+	}
+	if err := p.parseMetadataHeader(); err != nil {
+		return Geometry{}, err
+	}
+	if p.hasExt {
+		if err := p.parseExtendedPrecision(); err != nil {
+			return Geometry{}, err
+		}
+	}
+	if p.hasSize {
+		if err := p.parseSize(); err != nil {
+			return Geometry{}, err
+		}
+	}
+	if p.hasBBox {
+		if err := p.parseBBox(); err != nil {
+			return Geometry{}, err
+		}
+	}
+
+	switch p.kind {
+	case twkbTypePoint:
+		pt, err := p.parsePoint()
+		if err != nil {
+			return Geometry{}, err
+		}
+		return pt.AsGeometry(), nil
+	case twkbTypeLineString:
+		ls, err := p.parseLineString()
+		if err != nil {
+			return Geometry{}, err
+		}
+		return ls.AsGeometry(), nil
+	case twkbTypePolygon:
+		poly, err := p.parsePolygon()
+		if err != nil {
+			return Geometry{}, err
+		}
+		return poly.AsGeometry(), nil
+	case twkbTypeMultiPoint:
+		mp, err := p.parseMultiPoint()
+		if err != nil {
+			return Geometry{}, err
+		}
+		return mp.AsGeometry(), nil
+	case twkbTypeMultiLineString:
+		ml, err := p.parseMultiLineString()
+		if err != nil {
+			return Geometry{}, err
+		}
+		return ml.AsGeometry(), nil
+	case twkbTypeMultiPolygon:
+		mp, err := p.parseMultiPolygon()
+		if err != nil {
+			return Geometry{}, err
+		}
+		return mp.AsGeometry(), nil
+	case twkbTypeGeometryCollection:
+		gc, err := p.parseGeometryCollection()
+		if err != nil {
+			return Geometry{}, err
+		}
+		return gc.AsGeometry(), nil
+	}
+
+	return Geometry{}, p.parserError("cannot unmarshal unsupported geometry type")
+}
+
+func (p *TWKBParser) parseTypeAndPrecision() error {
+	if len(p.twkb) <= p.pos {
+		return p.parserError("cannot unmarshal empty buffer")
+	}
+	typeprec := p.twkb[p.pos]
+	p.pos++
+
+	p.kind = int(typeprec & 0x0f)
+	prec := int(DecodeZigZagInt32(uint32(typeprec) >> 4))
+	p.precX = prec
+	p.precY = prec
+
+	p.scalings[0] = math.Pow10(-p.precX)
+	p.scalings[1] = math.Pow10(-p.precY)
+	return nil
+}
+
+func (p *TWKBParser) parseMetadataHeader() error {
+	if len(p.twkb) <= p.pos {
+		return p.parserError("cannot unmarshal lacking metadata header")
+	}
+	metaheader := p.twkb[p.pos]
+	p.pos++
+
+	p.hasBBox = (metaheader & 1) != 0
+	p.hasSize = (metaheader & 2) != 0
+	p.hasIDs = (metaheader & 4) != 0
+	p.hasExt = (metaheader & 8) != 0
+	p.isEmpty = (metaheader & 16) != 0
+	return nil
+}
+
+func (p *TWKBParser) parseExtendedPrecision() error {
+	if len(p.twkb) <= p.pos {
+		return p.parserError("cannot unmarshal lacking extended pecision byte")
```

Comment:
```suggestion
		return p.parserError("cannot unmarshal: lacking extended precision byte")
```

Reply:
nit: I suggest adding a `:` after each `cannot unmarshal` in error messages (same below) as I feel we're annotating two error messages here and it helps with reading the error message:
- The fact that unmarshalling failed.
- The reason why unmarshalling failed.

Reply:
Good point, I'll make the error messages a bit clearer.